core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala

```diff
@@ -0,0 +1,746 @@
+      createTempBucket,
+      issueWorkspaceBucketTransferJob,
+      deleteWorkspaceBucket,
+      createFinalWorkspaceBucket,
+      issueTmpBucketTransferJob,
+      deleteTemporaryBucket,
+      updateWorkspaceRecord
+    )
+      .reverse
+      .traverse_(runStep)
+
+
+  // Sequence the action and return an empty MigrateAction if the action succeeded
+  def runStep(action: MigrateAction[Unit]): MigrateAction[Unit] =
+    action.mapF(optionT => OptionT(optionT.value.as(().some)))
+
+
+  final def startMigration: MigrateAction[Unit] =
+    for {
+      migration <- findMigration(_.started.isEmpty)
+      now <- nowTimestamp
+      _ <- inTransaction { _ =>
+        workspaceMigrations
+          .filter(_.id === migration.id)
+          .map(_.started)
+          .update(now.some)
+      }
+    } yield ()
+
+
+  final def claimAndConfigureGoogleProject: MigrateAction[Unit] =
+    for {
+      migration <- findMigration { m =>
+        m.started.isDefined && m.newGoogleProjectConfigured.isEmpty
+      }
+
+      workspace <- getWorkspace(migration.workspaceId)
+      errorDetails = Map(
+        ("migrationId" -> migration.id),
+        ("workspaceId" -> workspace.workspaceId)
+      )
+
+      _ <- MigrateAction.raiseWhen(workspace.billingAccountErrorMessage.isDefined) {
+        WorkspaceMigrationException(
+          message = "Workspace migration failed: billing account error exists on workspace",
+          data = errorDetails + ("billingAccountErrorMessage" -> workspace.billingAccountErrorMessage.get)
+        )
+      }
+
+      workspaceBillingAccount <- MigrateAction.liftIO(IO {
+        workspace.currentBillingAccountOnGoogleProject.getOrElse(
+          throw WorkspaceMigrationException(
+            message = "Workspace migration failed: no billing account on workspace",
+            data = errorDetails
+          )
+        )
+      })
+
+      // Safe to assume that this exists if the workspace exists
+      billingProject <- inTransactionT { _
+        .rawlsBillingProjectQuery
+        .load(RawlsBillingProjectName(workspace.namespace))
+      }
+
+      billingProjectBillingAccount <- MigrateAction.liftIO(IO {
+        billingProject.billingAccount.getOrElse(
+          throw WorkspaceMigrationException(
+            message = "Workspace migration failed: no billing account on billing project",
+            data = errorDetails + ("billingProject" -> billingProject.projectName)
+          )
+        )
+      })
+
+      _ <- MigrateAction.raiseWhen(billingProject.invalidBillingAccount) {
+        WorkspaceMigrationException(
+          message = "The migration failed: invalid billing account on billing project",
+          data = errorDetails ++ Map(
+            ("billingProject" -> billingProject.projectName),
+            ("billingProjectBillingAccount" -> billingProjectBillingAccount)
+          )
+        )
+      }
+
+      _ <- MigrateAction.raiseWhen(workspaceBillingAccount != billingProjectBillingAccount) {
```

Comment:
Can you give a code example? I don't quite follow your suggestion.

Reply:
As I am typing out the code, I think this might be more complexity than its worth, but here goes.

So, for example. (And this is kind of pseudocode-y?)
```
migrationPrerequisites = List.of(
  {
    predicate: (workspace, billingProject) => getWorkspaceBillingAccount(workspace) != getBillingProjectBillingAccount(billingProject)
    message: "Workspace migration failed: billing account on workspace differs from billing account on billing project"
    detailKeys: List.of("migrationId", "workspaceId", "workspaceBillingAccount", "billingProject", "billingProjectBillingAccount")
  },
  {
    predicate: (workspace, billingProject) => billingProject.invalidBillingAccount
    message: "Workspace migration failed: no billing account on billing project"
    detailKeys: List.of("migrationId", "workspaceId", "billingProject")
  }
)
```

What I think this would do is abstract out the business logic into this list. Then, our error handling code turns into:
```
migrationPrerequisites.map(migrationPrerequisiteObject => MigrateAction.raiseWhen(migrationPrerequisiteObject) {
  WorkspaceMigrationException(message = migrationPrerequisiteObject.message, data = errorDetailsFullList.filter(errorDetailOption => migrationPrerequisiteObject.detailKeys.contains(errorDetailOption.key)))
}

I think the error keys and ensuring that the args to the predicate are always the same is where the complexity builds, but the thought is that the error object construction be done once, so if we change the way we want to throw errors, we only need to change the `migrationPrerequisites.map` code block once, as opposed to for each of the prerequisites. This would help avoid situations like the one where there was one set not in the `raiseWhen`, and I was confused, but its just that it was missed.

Reply:
I like the idea, I just think that's adding too much complexity for this. I agree that all the checking makes it harder to understand what the funciton is meant to do. I think I'll pull this error handing out and group into their own functions to clean things up

Reply:
I agree that its a little over-engineered. The predicate stuff was particularly hard to figure out in a way that as I was writing it up seemed rough. I was mostly just noticing that much of the structure seemed to echo other lines around it, so was curious if we could break it into a map or a function in order to avoid the code duplication, but reducing the code duplication isn't worth adding complexity in my book.

Reply:
OK I tried many things - I tried a reduce, I tried to write simple header functions, I tired using `StateT` - it all leads to code explosion or increased complexity. I'm going to leave this.